File: algorithm/top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {

public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        map<int, int> freq;
        vector<vector<int>> counts(nums.size() + 1);
        vector<int> res;
        
        for (auto n: nums) {
            try {
                freq[n]++;
            } catch (out_of_range) {
                freq[n] = 1;
            }
        }

        for (auto it = freq.begin(); it != freq.end(); it++)
            counts[it->second].push_back(it->first);
        
        for (auto row = counts.rbegin(); row != counts.rend() && k > 0; row++)
            for (auto col = row->begin(); col != row->end() && k > 0; col++, k--)
                res.push_back(*col);

        return res;
    }
};

File: algorithm/top-k-frequent-elements/top-k-frequent-elements.go
package top_k_frequent_elements

import (
	"sort"
)

type node struct {
	num   int
	count int
}

func topKFrequent(nums []int, k int) []int {
	freq := make(map[int]int)

	for _, num := range nums {
		freq[num]++
	}

	nodes := make([]node, 0, len(freq))

	for num, count := range freq {
		nodes = append(nodes, node{num, count})
	}

	sort.Slice(nodes, func(i, j int) bool {
		return nodes[i].count > nodes[j].count
	})

	res := make([]int, 0, k)

	for i := 0; i < k; i++ {
		res = append(res, nodes[i].num)
	}

	return res
}
